File: 2023/c/1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <Windows.h>

int main()
{

  char *nums[] = {"1", "2", "3", "4", "5", "6", "7", "8", "9", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
  FILE *ptr;
  char ch;
  ptr = fopen("../2023-1-input.txt", "r");
  if (ptr == NULL)
  {
    printf("Error!");
    exit(1);
  }

  LARGE_INTEGER frequency, start, stop;

  // Get the frequency of the performance counter
  QueryPerformanceFrequency(&frequency);

  // Get the starting time
  QueryPerformanceCounter(&start);

  // Your function or code here
  char line[100];
  int result = 0;
  int secondResult = 0;
  while (fgets(line, sizeof(line), ptr) != NULL)
  {
    {
      int lineLength = strlen(line);
      char numStr[3];
      for (int i = 0; i < lineLength; i++)
      {
        char currentChar = line[i];
        if (isdigit(currentChar))
        {
          numStr[0] = currentChar;
          break;
        }
      }
      for (int i = lineLength - 1; i >= 0; i--)
      {
        char currentChar = line[i];
        if (isdigit(currentChar))
        {
          numStr[1] = currentChar;
          break;
        }
      }
      numStr[2] = '\0';
      int num = atoi(numStr);
      result += num;
    }
  }

  QueryPerformanceCounter(&stop);

  long long elapsed_us = (stop.QuadPart - start.QuadPart) * 1e6 / frequency.QuadPart;
  printf("\nElapsed time: %lld microseconds\n", elapsed_us);

  // long long elapsed_ns = (stop.QuadPart - start.QuadPart) * 1e9 / frequency.QuadPart;
  // printf("\nElapsed time: %lld nanoseconds\n", elapsed_ns);

  // long long elapsed_ms = (stop.QuadPart - start.QuadPart) * 1e3 / frequency.QuadPart;

  // printf("Elapsed time: %lld milliseconds\n", elapsed_ms);
  printf("Result: %d\n", result);
  fclose(ptr);
  return 0;
}

File: 2023/1.js
import { getAdventOfCodeData } from '../node-utils.js';
const start = performance.now();
const input = await getAdventOfCodeData(2023, 1);

let firstResult = 0;
let secondResult = 0;
const nums = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'];
const numStrMap = {
  'one': '1',
  'two': '2',
  'three': '3',
  'four': '4',
  'five': '5',
  'six': '6',
  'seven': '7',
  'eight': '8',
  'nine': '9',
};

input.split('\n').forEach((line) => {
  {
    // first part
    let numStr = '';
    for (let i = 0; i < line.length; i++) {
      const char = line[i];
      if (!isNaN(char)) {
        numStr += char;
        break;
      }
    }
    for (let i = line.length - 1; i >= 0; i--) {
      const char = line[i];
      if (!isNaN(char)) {
        numStr += char;
        break;
      }
    }
    firstResult += parseInt(numStr);
  }
  {
    // second part
    const numsIndexes = [];
    nums.forEach((numStr) => {
      const index = line.indexOf(numStr);
      if (index !== -1) {
        numsIndexes.push({index, value: numStr});
      }
      const lastIndex = line.lastIndexOf(numStr);
      if (lastIndex !== -1 && lastIndex !== index) {
        numsIndexes.push({index: lastIndex, value: numStr});
      }
    });
    numsIndexes.sort((a, b) => a.index - b.index);
    let twoNumStr = '';
    const firstNumStr = numsIndexes[0].value;
    const lastNumStr = numsIndexes[numsIndexes.length - 1].value;
    if (firstNumStr.length === 1) {
      twoNumStr = firstNumStr;
    } else {
      twoNumStr = numStrMap[firstNumStr];
    }
    if (lastNumStr.length === 1) {
      twoNumStr += lastNumStr;
    } else {
      twoNumStr += numStrMap[lastNumStr];
    }
    secondResult += parseInt(twoNumStr);
  }
});
const end = performance.now();

console.log('first result', firstResult);
console.log('secondResult', secondResult)
console.log('time taken', end - start, 'ms');